File: sdk/python/src/fhe_agent_shield/credentials.py
from typing import List, Dict
from .config import FHEConfig
from .contracts import AgentVault
from .types import CredentialResult, CredentialEntry


class FHECredentialStore:
    def __init__(self, config: FHEConfig):
        self.config = config
        self.vault = AgentVault(config.w3, config.account, config.contracts.agent_vault)
        self._credentials: Dict[str, str] = {}

    def store(self, key: str, value: str) -> str:
        encrypted = self._encrypt_value(value)
        tx_hash = self.vault.store_credential(self.config.address, encrypted)
        self._credentials[key] = tx_hash
        return tx_hash

    def retrieve(self, key: str, approvers: List[str]) -> str:
        handle = self._credentials.get(key)
        if not handle:
            raise ValueError(f"Credential '{key}' not found")
        handle_bytes = bytes.fromhex(handle) if isinstance(handle, str) else handle
        return self.vault.retrieve_credential(handle_bytes)

    def list(self) -> List[CredentialEntry]:
        return [
            CredentialEntry(key=key, handle=handle, threshold=1)
            for key, handle in self._credentials.items()
        ]

    def _encrypt_value(self, value: str) -> bytes:
        return bytes.fromhex(value.encode().hex().ljust(64, "0"))

File: sdk/moltbook/src/credentials.ts
/**
 * FHE Credential Manager for Moltbook
 * 
 * Manages FHE-encrypted credentials for Moltbook agents.
 */

import type { FHEConfig, Credential } from './types.js';

interface CredentialConfig {
  network: 'ethereum-sepolia' | 'arbitrum-sepolia';
  privateKey: `0x${string}`;
  rpcUrl: string;
  contracts: {
    agentVault: `0x${string}`;
    agentMemory: `0x${string}`;
  };
}

export class FHECredentialManager {
  private config: CredentialConfig;
  private credentials: Map<string, Credential> = new Map();

  constructor(config: CredentialConfig) {
    this.config = config;
  }

  /**
   * Initialize FHE vault for an agent
   */
  async initialize(agentId: string): Promise<void> {
    // In production, this would call AgentVault.initialize() on Fhenix
    console.log(`[FHE] Initializing vault for agent ${agentId}`);
  }

  /**
   * Store a credential with FHE encryption
   */
  async store(key: string, value: string, threshold = 2): Promise<string> {
    // In production: call AgentVault.storeCredential() with FHE ciphertext
    // Mock implementation
    const encryptedKey = this.fheEncrypt(key);
    const encryptedValue = this.fheEncrypt(value);

    const credential: Credential = {
      id: `cred_${key}_${Date.now()}`,
      encryptedKey,
      encryptedValue,
      owner: '0x' + '0'.repeat(40),
      threshold,
    };

    this.credentials.set(key, credential);
    return credential.id;
  }

  /**
   * Retrieve a credential with threshold authorization
   */
  async get(key: string, permits = 2): Promise<string | null> {
    const cred = this.credentials.get(key);
    if (!cred) return null;

    if (permits < cred.threshold) {
      throw new Error(`Insufficient permits: need ${cred.threshold}, got ${permits}`);
    }

    // In production: threshold decryption via Fhenix CoFHE
    // Mock: return mock decrypted value
    return `decrypted_${key}`;
  }

  /**
   * List all credentials
   */
  list(): Array<{ id: string; key: string; threshold: number }> {
    return Array.from(this.credentials.entries()).map(([key, cred]) => ({
      id: cred.id,
      key,
      threshold: cred.threshold,
    }));
  }

  /**
   * Delete a credential
   */
  async delete(key: string): Promise<void> {
    this.credentials.delete(key);
  }

  /**
   * FHE encrypt data (mock implementation)
   * Production uses Fhenix CoFHE
   */
  private fheEncrypt(data: string): `0x${string}` {
    // Mock: return keccak256 of data as hex
    const hash = data.split('').reduce((acc, c) => {
      return ((acc << 5) - acc + c.charCodeAt(0)) | 0;
    }, 0);
    return `0x${Math.abs(hash).toString(16).padStart(64, '0')}`;
  }
}
